### ISSUE TITLE
Willing to collaborate on pushing forward this project.

### ISSUE THREAD

**Issue by yhyu13** [jprothero/MCTSnet#1]
Hi, @jprothero 

It's great to see another deep learning enthusiast. Greetings.
You can call me Yohan, I am a junior year Math student at University of illinois. You can email me `hangyu5@illinois.edu`

I would like to work on the idea of MCTSnet this summer break. I personally think DeepMind has a made a great deal on this idea but the published paper was rejected by ICLR2018 because the success on Sokoban was not considered a significant result as its average branch factor is 12 for a 20x20 maze (DeepMind failed to detail its experiment setup as suggested on OpenReview). 

I hope we can expand it to supervised training against Go professional player's move. I had experience to replicate AlphaGo Zero's supervised learning result ([here](https://github.com/yhyu13/AlphaGOZero-python-tensorflow)). I used Nvidia docker with AWS Volta instance and around 3 hours to finish training for a 20 layers residual network suggested in the AlphaGo Zero paper. Ultimately, I would like to give MCTSnet a try on minRTS from Facebook ELF to see if it can beat built-in AI ([paper](https://arxiv.org/pdf/1707.01067.pdf) page 7)

Also, I tried to run ```python main.py``` but it seems like your have a newer version that include `newMCTS` which is not present in the current repository. I wonder what is your experience on training MCTSnet on connecte4, does it take long? And how much resource does it take? I am willing to see your update so that we can proceed further. 

Your code looks neat, I think you are a better coder than me /:D.

**Comment by jprothero** [comment]
Hi @yhyu13, 

Your work replicating supervised AlphaGo Zero looks very well done. I've researched AlphaGo / AlphaGo Zero a bit too. Thank you for the complement on my code, I still have a lot to learn :)

I would be happy to collaborate together and try to do some better experiments with MCTSnet such as with Connect4, Go or even the miniRTS you mentioned. I think getting it working on Connect4 first would be a good step, but maybe you could work on adapting it to other games.

The repository is a bit messy right now. I was kind of in the middle of changing a lot of things then I took a break from working on this project. Since you're interested in it I can fix it up and try to get it in a working state.

About the resource requirements for the project, I currently do not have GPU so it's hard to fully test the code. I plan to buy a new Nvidia GPU once the next generation comes out. If we get the code working we could rent a cloud GPU on Paperspace or AWS.

Were you mainly interested in replicating the MCTSnet paper and trying it out on a new domain, or are you interested in altering it a bit and kind of making it our own way? I kind of would like to take the core ideas from the paper and try to simplify it and maybe make it a bit better if possible. Let me know in what ways you're interested in collaborating.

For now I'll try to fix it up and get it in a cleaner, working state, and then I can send you a message to let you know.

**Comment by yhyu13** [comment]
Hi, @jprothero 

Reproducibility of MCTSnet paper is rather challenging. The author of MCTSnet used sokoban implementation designed by Victor Valdes (deepmind engineer) originally programmed for I2A (another DeepMind RL paper) . In the appendix of MCTSnet paper, it mentions how the supervised label is generated by MCTS with leaf node evaluated by a pre-trained network. It also mentions the layout of Sokoban game, and its level design is the same as I2A paper. 

The main challenge is to figure out how the pre-trained network is trained and make our own version of DeepMind's sokoban game generator, both of them make replication not easy to follow.  Even though reproducibility is not an issue according to the  ICLR2018 rejection [comment](https://openreview.net/forum?id=r1TA9ZbA-), I believe it would a problem if DeepMind doesn't release its sokoban program. 

Judging from that, I will follow what your are doing and apply the core idea in broader tasks.

**Comment by jprothero** [comment]
@yhyu13 

Okay, that sounds like a good plan. I will fix up the project and send you a message once that's done.

**Comment by jprothero** [comment]
@yhyu13 
I pushed some changes. It's slow at the moment and I'm pretty sure it has some bugs but you can probably start playing around with it if you want to.

EDIT:
It has a lot more bugs than I realized, I'll probably need another day or two to get it 100%

**Comment by yhyu13** [comment]
@jprothero 

I appreciate your dedicated updates. I wonder if you use anaconda or pip to manage python dependencies (yaml or requirement text),  it seems to me you are using py3.6 or higher because I noticed f-string. Let me know what is your development environment with all the essential packages I need to know.

**Comment by jprothero** [comment]
@yhyu13 

I'm just using pip for now. I added a requirements.txt in the repository.

I use Python 3.6 I think. I'm pretty sure the f-string thing is the only recent python thing that I use. I can switch it out since it's pretty minor.

**Comment by yhyu13** [comment]
Hi, @jprothero 

I am fine with py3.6 as f-string is aesthetic. Since I often write code for scientists who don't use python, I find many of them use py3.5 which is more popular, we can stick with py3.6 in this project as along as you mention it explicitly. Also, I am prune to use a python logging package 'daiquiri' a lot since it prints datetime with cogent level of messages, you will find more when I start contributing to this project, just let  you know.

**Comment by yhyu13** [comment]
@jprothero 

I found qrnn requires tensorflow, I am not sure where it is used in your code. I haven't use this package before, and I found there is [pytorch implementation](https://github.com/salesforce/pytorch-qrnn), maybe you will consider use that instead?

**Comment by jprothero** [comment]
@yhyu13 
I agree f-string is cleaner, but a lot of people have Python 3.5 like you said. I think we should try to keep it 3.5 compatible if possible. Daiquiri sounds interesting, we can add that in for sure.

I'm not sure about the QRNN thing, I might have put that in the repository just to experiment. I'm not using it at the moment, so I'll try to find it and take it out.

**Comment by yhyu13** [comment]
@jprothero 

Sure, py3.5 is cool to me. Also, I suggest to remove __pycache__ from the repository as it cause more merge conflict and it will generate anyway.

**Comment by jprothero** [comment]
@yhyu13 

Good catch, I added a .gitignore with some different things.

**Comment by jprothero** [comment]
@yhyu13 
I think everything is working now, free of any major bugs. It is suuuuuper slow though atm, so I'm going to try to fix that.

**Comment by jprothero** [comment]
@yhyu13 

I've been testing it a bit and it appears to be training correctly, the newly trained network appears in general to beat the existing best network. You could try tweaking the model_config.py file with different numbers and see if any of them improve performance, or modifying MCTSnet_model.py directly. I only have access to the free GPU from Google Colab, so I can't test using a larger network at the moment.

The next step I need to do is figuring out how to batch games together so that the GPU is being used more efficiently, but that might take a while.

EDIT: Just noticed another bug, might invalidate the training success.

**Comment by yhyu13** [comment]
Hi @jprothero

I started training right after you told me it worked. My experience matches what you said in the previous comment, just let you know the pace it runs on my machine: about 10s/it for NUM_SIMS=15 and 2 s /it for NUM_SIMS=2 (minimum simulation allowed in your program). I think it's fair enough.

I do have a couple ideas from facebook ELF where you start many games at once and feed states as batches into neural network, and update each game with corresponding action in the batched output. It is the method that helps ELF to run 16 minRTS games at once. I understand it in high level but I haven't programmed it yet.  In AlphaGo zero, we can easily parallelize the expansion of leaf nodes. Let me know if it's applicable in MCTSnet since it has memory statistics in addition to the board state.

Also, another idea examine MCTSnet is to let it play against an MinMax AI, I can work on that if you allow me to be a contributor (I will start a new branch).

**Comment by jprothero** [comment]
@yhyu13 

The parallel ELF idea sounds good, I think I tried something similar to that before. I can try to add that. We can try to figure out the multiprocessing together. I think the main bottleneck is lack of GPU batching right now though.

Having a MinMax baseline would be good for sure. I'll add you as a contributor.

**Comment by yhyu13** [comment]
Hi @jprothero

I've added a file called MinMax.py and its test function in a folder called `test`. You can run test just by `python3 -m pytest`. If that works fine, you can play around with it by `python MinMax`. It runs fast and it show some intelligence. 

I am trying to make a class called MCTSnetPlayerMixin, where an MCTSnet agent can be evaluated against another AI (hopefully we can make a interface for us to play as well). I still trying to wrap my head around how to integrate MinMax into your program, so it will take a day or two.

**Comment by jprothero** [comment]
Hi @yhyu13 

Cool, awesome work. Just a heads up I'm going to be busy for a couple weeks, so I won't be able to spend as much time on the project during that time. Let know if you have any questions about the code and thanks for helping out. Hopefully we can make this project great together!

**Comment by yhyu13** [comment]
Hi, @jprothero 

Just let you know I was having a short break last week and I am back to work. I grasped your idea and hopefully you will see my commits on MCTSnet play against baseline in one or two days. Thanks making this project happen!

**Comment by jprothero** [comment]
Hey, @yhyu13 

Great, thanks so much for contributing. I still have some other things I need prioritize for a while, but I hope to help with the project in a couple of weeks. Please let me know if you have any questions or feature requests and I'll try to help. Thanks again.

**Comment by yhyu13** [comment]
Hi, @jprothero 

I've made a pull request to your master branch the other day, just let you know. It would be great if you can review my code very quickly. 

I've trained a self-play agent overnight (400 iterations) and make it play against minmax (search depth 4)for 10 rounds, but it lose 10 rounds totally. I would not say it's upset because the agent only has self play 4000 self-play games experience, I will continue training it.

**Comment by jprothero** [comment]
@yhyu13 

I looked over your code, it looks pretty good. This project is pretty informal so as long as things work I'm fine with whatever changes. There may be a bug in the code that is causing the MCTSnet to not work quite right, or just the neural net model isn't that great. When I get a chance to work on this project again (hopefully this weekend), I think I will try to change the style of the MCTSnet to be closer to the original paper and we can see how that works out.

**Comment by jprothero** [comment]
Hey @yhyu13 

I tried out a change to the model and so far it seems to be working pretty well. Could you test it out against the MinMax and let me know how it does?

**Comment by yhyu13** [comment]
@jprothero 

Sorry for the late reply, I was just started applying your changes. I will let you know how well it works after a whole day of training by tomorrow. Thanks again for your work!

**Comment by jprothero** [comment]
@yhyu13 

Hey, sorry about that. I accidentally didn't push the changes to master. It is ready for testing now. Thanks for testing it out, I appreciate it a lot!

**Comment by yhyu13** [comment]
Hi @jprothero 

Don't worry, I started over right after you latest update.

**Comment by yhyu13** [comment]
Hi, @jprothero 

With 600 iterations of training with 12 simulations per move, the MCTSnet still could not outplay a Minmax agent with 4 search depth (0-10). So I made two changes for further testing, one is the number of simulations per move, the other is the search depth. I found that with 30 simulations and 1 search depth, the score becomes (3-7). Holding the same number of simulations with 2 search spit out (0-10) again. 

I believe the MCTSnet achieve some intelligence from self-play, but it is not strong enough. So there are two directions I will be improving, one is the to parallelize the self-play methods (I believe it shall be done inside the while loop of main.py), the other is to enlarge the size of the neural net. Now I have other things to concern, I will carry on them tomorrow.

**Comment by jprothero** [comment]
@yhyu13

Interesting results, thanks for testing the net.

I'm not too sure what other tweaks to the network to try. I might look into using meta learning or try using some of the techniques of the playing atari with 6 neurons paper. That might take a little while to get done though. 

If you want you could consider trying to replicate the system from the MCTSnet paper.

Thanks again for all the hard work!

**Comment by jprothero** [comment]
Hi @yhyu13 

Just wanted to give you a heads up that I'm starting a new job soon and I won't have too much time to work on this project for a while. Please let me know if you need help with anything, and I'll try to give an answer promptly and help as much as I can. Thanks for all the help so far, hopefully we can eventually make some good progress with this project.

**Comment by yhyu13** [comment]
@jprothero 

I just finished GRE yesterday so that I can make more progress on the two tasks I mentioned earlier (larger network and parallel simulations). I will push commits once it's bug free. Thanks for your effort too and congratulate on the new job!

**Comment by jprothero** [comment]
@yhyu13 

Thank you! Congrats on finishing the GRE.
